### PR TITLE
Breaking: Update nom to v7. Use lexical for writing floating point values (changing formatting output)

### PR DESCRIPTION
**Pull request by genos**
# Bottom Line Up Front

In #71, I pulled in [`ryu`](https://docs.rs/ryu/latest/ryu/) to format floating point values in `format_complex()`, not realizing that `quil-rs` already depends upon [`lexical`](https://docs.rs/lexical/latest/lexical/). It turns out that `lexical` formats things very nicely (including an option to strip trailing zeros, which I didn't find in `ryu`), can be configured to output scientific notation with the same bounds as `ryu`, and looks faster in benchmarking (see below).

# Equivalent output to `ryu`

The following passes as many tests as I care to run locally

<details>
  <summary>Testing with proptest to ensure equivalence</summary>

```rust
use num_complex::Complex64;
use std::num::NonZeroI32;

pub fn using_lexical(value: Complex64) -> String {
    const FORMAT: u128 = lexical::format::STANDARD;
    let options = lexical::WriteFloatOptions::builder()
        .negative_exponent_break(NonZeroI32::new(-5))
        .positive_exponent_break(NonZeroI32::new(15))
        //.trim_floats(true)  // We _want_ this but it messes up testing against ryu.
        .build()
        .unwrap();
    if value.re == 0f64 && value.im == 0f64 {
        "0".to_owned()
    } else if value.im == 0f64 {
        lexical::to_string_with_options::<_, FORMAT>(value.re, &options)
    } else if value.re == 0f64 {
        lexical::to_string_with_options::<_, FORMAT>(value.im, &options) + "i"
    } else {
        let mut out = lexical::to_string_with_options::<_, FORMAT>(value.re, &options);
        if value.im > 0f64 {
            out.push('+')
        }
        out.push_str(&lexical::to_string_with_options::<_, FORMAT>(
            value.im, &options,
        ));
        out.push('i');
        out
    }
}

pub fn using_ryu(value: Complex64) -> String {
    if value.re == 0f64 && value.im == 0f64 {
        "0".to_owned()
    } else if value.im == 0f64 {
        ryu::Buffer::new().format(value.re).to_owned()
    } else if value.re == 0f64 {
        ryu::Buffer::new().format(value.im).to_owned() + "i"
    } else {
        let mut buf = ryu::Buffer::new();
        let op = if value.im > 0f64 { "+" } else { "" };
        buf.format(value.re).to_owned() + op + buf.format(value.im) + "i"
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;

    fn arb_complex64() -> impl Strategy<Value = Complex64> {
        any::<(f64, f64)>().prop_map(|(re, im)| Complex64::new(re, im))
    }

    proptest! {
        #[test]
        fn equivalent_but_for_trailing_zeros(value in arb_complex64()) {
            assert_eq!(using_lexical(value), using_ryu(value));
        }
    }
}
```
</details>

# Still Speedy

With the following [`criterion`](https://bheisler.github.io/criterion.rs/criterion/) benchmarking,

<details>
<summary>Benchmarking with criterion</summary>

```rust
use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion};
use format_comparison::{using_lexical, using_ryu};
use num_complex::Complex64;

fn bench_formats(c: &mut Criterion) {
    let mut group = c.benchmark_group("Formatting");
    for size in 0..10 {
        let e = 10f64.powi(size);
        let re = rand::random::<f64>() * e;
        let im = rand::random::<f64>() * e;
        let value = Complex64::new(re, im);
        group.bench_with_input(BenchmarkId::new("using_lexical", size), &size, |b, _i| {
            b.iter(|| using_lexical(black_box(value)))
        });
        group.bench_with_input(BenchmarkId::new("using_ryu", size), &size, |b, _i| {
            b.iter(|| using_ryu(black_box(value)))
        });
    }
    group.finish();
}

criterion_group!(benches, bench_formats);
criterion_main!(benches);
```

</details>

we see that `lexical` looks faster (and perhaps more consistent):

violin
lines

### CODE REVIEW

**Review by dbanty — CHANGES_REQUESTED** [review]
😭 but I only just updated my tests elsewhere to add those trailing 0s.

In all seriousness though, this looks great! Are we parsing any numbers elsewhere that could also benefit from using `lexical`?

**Comment by genos** [comment]
Looks like this is failing a license `deny` check, but `lexical` has both [Apache 2.0](https://github.com/Alexhuszagh/rust-lexical/blob/main/LICENSE-APACHE) and [MIT](https://github.com/Alexhuszagh/rust-lexical/blob/main/LICENSE-MIT) licensing. @dbanty halp plz?

File: src/expression.rs [hunk 1]
```diff
@@ -435,16 +436,29 @@ impl<'a> FromStr for Expression {
 /// - When both are non-zero, show with the correct operator in between
 #[inline(always)]
 fn format_complex(value: &Complex64) -> String {
+    const FORMAT: u128 = lexical::format::STANDARD;
+    let options = lexical::WriteFloatOptions::builder()
+        .negative_exponent_break(NonZeroI32::new(-5))
+        .positive_exponent_break(NonZeroI32::new(15))
+        .trim_floats(true)
+        .build()
+        .unwrap();
```
[the remark this answers is not in the document]
Comment (genos) [reply src/expression.rs:445]: Excellent points, thanks. It's not guaranteed in the library to never panic, but I think what little options we're setting here are in fact safe, so I'll document that. As for `const`-ness, can an `unwrap` be `const`?

**Comment by dbanty** [comment]
The license complaint is about a license we’re allowing but not using—that might just be a warning.

the failure about this package is because of a duplicate. If there’s not a newer version of nom that uses the newer lexical than maybe we should stick with the older lexical we had (unless there’s something in v6 we need).

the other option is to tell deny to allow this duplicate

File: src/expression.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (genos) [reply src/expression.rs:445]: Ah, there's also an `unsafe build_unchecked()`, with explicit documentation about its safety (plus I looked through the source). Update incoming.

**Comment by genos** [comment]
adee223 is labeled as **BREAKING** since it updates `nom`, and we return some `nom` types in parsing.

**Review by dbanty — APPROVED** [review]
Approved! I also changed the title so that if we squash the breaking nature is captured properly.

**Review by nilslice — APPROVED** [review]
LGTM!

**Pull request merged**